_posts/blog/2016-09-06-model-view-viewmodel.md

```diff
@@ -0,0 +1,181 @@
+
+### View
+
+The View is the actual user interface in the app. It can be an `Activity`, a `Fragment` or any custom Android `View`. For `Activities` and `Fragment`s, we are binding and unbinding from the event sources on `onResume()` and `onPause()`.
+
+{% highlight java %}
+    private final CompositeSubscription mSubscription = new CompositeSubscription();
+
+    @Override
+    public void onResume() {
+        super.onResume();
+        mSubscription.add(mViewModel.getSomeData()
+                         .observeOn(AndroidSchedulers.mainThread())
+                         .subscribe(this::updateView,
+                                    this::handleError));
+    }
+
+    @Override
+    public void onPause() {
+        mSubscription.clear();
+        super.onPause();
+    }
+{% endhighlight %}
+
+In case the MVVM View is a custom Android `View`, the binding is done in the constructor. To ensure that the subscription is not preserved, leading to possible memory leaks, the unbinding happens in `onDetachedFromWindow`.
+
+{% highlight java %}
+    private final CompositeSubscription mSubscription = new CompositeSubscription();
+
+    public MyView(Context context, MyViewModel viewModel) {
+        ...
+        mSubscription.add(mViewModel.getSomeData()
+                         .observeOn(AndroidSchedulers.mainThread())
+                         .subscribe(this::updateView,
+                                    this::handleError));
+    }
+
+    @Override
+    public void onDetachedFromWindow() {
+        mSubscription.clear();
+        super.onDetachedFromWindow();
+    }
+}
+{% endhighlight %}
+
+## Testability Of The Model-View-ViewModel Classes
+
+One of the main reasons why we love the Model-View-ViewModel pattern is its high degree of testability.
+
+### DataModel
+
+The inversion of control pattern, heavily applied in our code, and the lack of any Android classes, make the DataModel easy to unit test.
+
+### ViewModel
+
+The ViewModel exposes streams of data to whoever needs to consume it. We see the Views and the unit tests as two different types of data consumers.
+The ViewModel is completely separated from the UI or any Android classes, therefore straightforward to unit test.
+
+Considering the following example where the ViewModel just exposes some data from the DataModel:
+
+{% highlight java %}
+public class ViewModel {
+    @NonNull
+    private final IDataModel mDataModel;
+
+    public ViewModel(@NonNull final IDataModel dataModel) {
+        mDataModel = dataModel;
+    }
+
+    @NonNull
+    public Observable<Data> getSomeData() {
+        return mDataModel.getSomeData();
+    }
+}
+{% endhighlight %}
+
+The tests for the ViewModel are easy to implement. With the help of Mockito, we are mocking the DataModel and we control the returned data for the methods used. Then, we make sure that when we subscribe to the `Observer` returned by `getSomeData()`, the expected data is emitted.  
```

Comment:
Probably overly repetitive with the "easy", consider finding a different way of saying (perhaps by stating what about it makes it easy to test)
> are easy to implement. 

Or you could just remove that sentence and lead with the Mockito bit. 

You could also refer to the technique of mocking and say "we use Mockito for this". 

Perhaps link to Mockito too.